### PR TITLE
MONGOCRYPT-384 AES-256 in CTR mode for OpenSSL

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

Add AES-256 in CTR mode with OpenSSL.

This leaves stub implementations for work to be done as follow up. Follow up work to this PR is as follows:
- Add AES-256 in CTR mode for CommonCrypto (macOS).
- Add AES-256 in CTR mode for Cryptography API: Next Generation (CNG) (Windows).
- Add a hook, similar to [mongocrypt_setopt_crypto_hooks](https://github.com/mongodb/libmongocrypt/blob/7b5e36155264699b62c88413efaf4c7a9638a230/src/mongocrypt.h.in#L1220-L1229) for drivers to add a callback for AES-256 in CTR mode. That is required by Python, Ruby, C#, Java, and Node. See [Which drivers implement libmongocrypt crypto callbacks?](https://docs.google.com/spreadsheets/d/1NKtmYEr_Zrf_eDATpfCBfWyLH8GC1vj075aSCLqrca0/edit#gid=0)

# Background & Motivation

CTR mode is a requirement for FLE 2.0. See [AEAD with CTR](https://docs.google.com/document/d/1eCU7R8Kjr-mdyz6eKvhNIDVmhyYQcAaLtTfHeK7a_vE/edit#).

### CODE REVIEW

**Review by chardan — APPROVED** [review]
LGTM-- there are a couple of wording changes you might want to consider.

File: src/crypto/libcrypto.c [hunk 1]
```diff
@@ -170,6 +179,30 @@ _native_crypto_aes_256_cbc_decrypt (const _mongocrypt_buffer_t *key,
    return ret;
 }
 
+bool
+_native_crypto_aes_256_cbc_encrypt (const _mongocrypt_buffer_t *key,
+                                    const _mongocrypt_buffer_t *iv,
+                                    const _mongocrypt_buffer_t *in,
+                                    _mongocrypt_buffer_t *out,
+                                    uint32_t *bytes_written,
+                                    mongocrypt_status_t *status)
```
Comment (vector-of-bool) [inline src/crypto/libcrypto.c:188]: This exact set of six parameters is repeated in many functions. Could they be encapsulated into a params structure to represent the parameters as named fields? That would be easier to pass around with accidentally transposing or confusing arguments.

File: src/crypto/libcrypto.c [hunk 2]
```diff
@@ -58,31 +58,36 @@ _native_crypto_init ()
...
-                                    const _mongocrypt_buffer_t *in,
-                                    _mongocrypt_buffer_t *out,
-                                    uint32_t *bytes_written,
-                                    mongocrypt_status_t *status)
+/* _encrypt_with_cipher encrypts @in with the OpenSSL cipher specified by
+ * @cipher.
+ * @key is the input key. @iv is the input IV.
+ * @out is the output ciphertext. @out must be allocated by the caller with
+ * enough room for the ciphertext.
+ * @bytes_written is the resulting bytes written to @out.
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/crypto/libcrypto.c:66]: Done.

File: src/crypto/libcrypto.c [hunk 3]
```diff
@@ -58,31 +58,36 @@ _native_crypto_init ()
...
+ * Returns false and sets @status on error. */
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/crypto/libcrypto.c:67]: It is required. Done.

File: src/crypto/libcrypto.c [hunk 4]
```diff
@@ -114,30 +119,34 @@ _native_crypto_aes_256_cbc_encrypt (const _mongocrypt_buffer_t *key,
...
-                                    const _mongocrypt_buffer_t *in,
-                                    _mongocrypt_buffer_t *out,
-                                    uint32_t *bytes_written,
-                                    mongocrypt_status_t *status)
+/* _decrypt_with_cipher decrypts @in with the OpenSSL cipher specified by
+ * @cipher.
+ * @key is the input key. @iv is the input IV.
+ * @out is the output plaintext. @out must be allocated by the caller with
+ * enough room for the plaintext.
+ * @bytes_written is the resulting bytes written to @out.
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/crypto/libcrypto.c:127]: Done.

File: src/crypto/libcrypto.c (hunk 1 above)
Comment (kevinAlbs) [reply src/crypto/libcrypto.c]: Added a `aes_256_args_t` type.

**Review by vector-of-bool — APPROVED** [review]
LGTM

**Pull request merged**